### PR TITLE
Improve effect analysis of bitshifts by non `Int` and `UInt` `Integer`s

### PR DESCRIPTION
**Pull request by oscardssmith**
this improves the effect analysis of things like `UInt32 >> Int32`

### CODE REVIEW

File: base/operators.jl [hunk 1]
```diff
@@ -641,7 +641,8 @@ See also [`>>`](@ref), [`>>>`](@ref), [`exp2`](@ref), [`ldexp`](@ref).
 """
 function <<(x::Integer, c::Integer)
     @inline
-    typemin(Int) <= c <= typemax(Int) && return x << (c % Int)
+    0 <= c <= typemax(UInt) && return x << (c % UInt)
+    -c <= typemax(UInt) && return x >> (-c % UInt)
     (x >= 0 || c >= 0) && return zero(x) << 0  # for type stability
     oftype(x, -1)
```
Comment (vtjnash) [inline base/operators.jl:647]:
```suggestion
    (x >= 0 || c >= 0) && return zero(x) << UInt(0)  # for type stability
    return oftype(x, -1) << UInt(0)
```

**Comment by giordano** [comment]
Can we have a more descriptive title than "mess with..."? Also, can the claim about improved effect analysis be backed up with tests?

**Comment by oscardssmith** [comment]
tests added.

File: test/int.jl [hunk 2]
```diff
@@ -199,6 +199,11 @@ end
                 @test val >> -scount === val << ucount
             end
         end
+        for T2 in Base.BitInteger_types
+            for op in (>>, <<, >>>)
+                Core.Compiler.is_total(Base.infer_effects(op, (T1, T2)))
```
Comment (thchr) [inline test/int.jl:204]:
```suggestion
                @test Core.Compiler.is_total(Base.infer_effects(op, (T1, T2)))
```

**Comment by oscardssmith** [comment]
adding backport-1.9 because this is needed for https://github.com/JuliaLang/julia/pull/47501 to be nothrow.

**Review by gbaraldi — APPROVED** [review]

File: base/operators.jl [hunk 3]
```diff
@@ -689,11 +689,11 @@ function >>(x::Integer, c::Integer)
     if c isa UInt
         throw(MethodError(>>, (x, c)))
     end
-    typemin(Int) <= c <= typemax(Int) && return x >> (c % Int)
+    0 <= c <= typemax(UInt) && return x >> (c % UInt)
```
[the remark this answers is not in the document]
Comment (oscardssmith) [reply base/operators.jl:692]: `c` could be a `BigInt` or any larger integer. Any of the `BitInteger`s will hit the methods in `int.jl`. These are just fallbacks for more complicated integer types.

**Comment by quinnj** [comment]
Did this perhaps break BitIntegers package? Seeing this on a dependent package nightly CI failure:
```
fatal: error thrown and no exception handler available.
MethodError(f=Base.:(<<), args=(0x00000000, 24), world=0x000000000000826b)
jl_method_error_bare at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:1942
jl_method_error at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:1960
jl_lookup_generic_ at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2688 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2703
Char at ./char.jl:162
AbstractChar at ./char.jl:48
parseint_iterate at ./parse.jl:51
parseint_preamble at ./parse.jl:58
tryparse_internal at ./parse.jl:107
#tryparse#509 at ./parse.jl:237
tryparse at ./parse.jl:235
repl_color at ./client.jl:17
error_color at ./client.jl:22
display_error at ./client.jl:110
unknown function (ip: 0x7f3b1177af46)
_jl_invoke at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2525 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2707
display_error at ./client.jl:114
unknown function (ip: 0x7f3b1177a852)
_jl_invoke at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2525 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2707
jl_apply at /cache/build/default-amdci5-6/julialang/julia-master/src/julia.h:1870 [inlined]
jl_f__call_latest at /cache/build/default-amdci5-6/julialang/julia-master/src/builtins.c:774
#invokelatest#2 at ./essentials.jl:816 [inlined]
invokelatest at ./essentials.jl:813 [inlined]
_start at ./client.jl:524
jfptr__start_[470](https://github.com/RelationalAI/rai-sdk-julia/actions/runs/3496167789/jobs/5854156556#step:3:491)94.clone_1 at /opt/hostedtoolcache/julia/nightly/x64/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2525 [inlined]
ijl_apply_generic at /cache/build/default-amdci5-6/julialang/julia-master/src/gf.c:2707
jl_apply at /cache/build/default-amdci5-6/julialang/julia-master/src/julia.h:1870 [inlined]
true_main at /cache/build/default-amdci5-6/julialang/julia-master/src/jlapi.c:573
jl_repl_entrypoint at /cache/build/default-amdci5-6/julialang/julia-master/src/jlapi.c:717
main at /cache/build/default-amdci5-6/julialang/julia-master/cli/loader_exe.c:58
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x401098)
ERROR: LoadError: Failed to precompile BitIntegers [c3b6d118-76ef-56ca-8cc7-ebb389d030a1] to /home/runner/.julia/compiled/v1.10/BitIntegers/jl_6QfzKk.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::IO, internal_stdout::IO, keep_loaded_modules::Bool)
    @ Base ./loading.jl:1820
```

**Comment by giordano** [comment]
Maybe the "mess with..." title was more accurate after all

**Comment by oscardssmith** [comment]
grr. I guess we have to revert this and make effect analysis smart enough to do this itself. @aviatesk does that seem possible to you?

**Comment by gbaraldi** [comment]
My pr has assume effects for both things currently, because with constprop LLVM figures out it can't throw anyway. So it's not a blocker or anything. It would just be nice

**Comment by aviatesk** [comment]
> grr. I guess we have to revert this and make effect analysis smart enough to do this itself.

What's the problem with this PR? Why not use `@assume_effects`?

**Comment by oscardssmith** [comment]
the hard part of this PR is that before it `UInt32 >> Int32` was hitting the method for `Integer>>Integer` which has a `throw` statement that would not be reached, but that the effects system couldn't tell wouldn't be reached.

**Comment by quinnj** [comment]
So......should we open a new issue to track that BitIntegers.jl is still broken (can't load) from this PR? Or should we revert? It's making testing latest w/ a lot packages not work.

**Comment by oscardssmith** [comment]
I guess I'll revert. It will make me sad, but I don't know of a better option.

**Pull request merged**